crates/compiler/src/error_listener.rs

```diff
@@ -52,3 +108,127 @@ pub enum DiagnosticSeverity {
     /// to fix, but are unlikely to cause problems.
     Info,
 }
+
+pub(crate) struct LexerErrorListener {
+    pub(crate) diagnostics: RefCell<Vec<Diagnostic>>,
+    file_name: String,
+}
+
+impl LexerErrorListener {
+    pub(crate) fn new(file_name: String) -> Self {
+        Self {
+            file_name,
+            diagnostics: Default::default(),
+        }
+    }
+}
+
+impl<'input, T: Recognizer<'input>> ErrorListener<'input, T> for LexerErrorListener {
+    fn syntax_error(
+        &self,
+        _recognizer: &T,
+        _offending_symbol: Option<&<T::TF as TokenFactory<'input>>::Inner>,
+        line: isize,
+        column: isize,
+        msg: &str,
+        _error: Option<&ANTLRError>,
+    ) {
+        let line = (line - 1) as usize;
```

Comment:
why the -1?

Reply:
Idk, took it from the original. I guess the error occurred in the penultimate line?